### PR TITLE
Add PipelineThread and AtomicBuffer classes.

### PR DESCRIPTION
**Pull request by Nicogene**
This PR introduces in `YARP_OS` two new classes: `PipelineThread` and `AtomicBuffer`.

These classes have been created time ago with @apaikan to optimize a pipeline implementation through multithreading.

I decided to import them in `YARP` because I need them in another project and I think that they can be useful for someone else.

I wrote with @aikolina a dox page with the description of these classes and a code example for using them.

Please review code.

### CODE REVIEW

File: doc/pipeline_thread/pipelineThread.dox [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+        ImageOf<PixelRgb> *image_yarp = image_port.read();
+
+        Mat image_cv = cvarrToMat(static_cast<IplImage*>(image_yarp->getIplImage()));
+        // convert to grayscale
+        cvtColor(image_cv, image_cv, CV_RGB2BGR);
+        cvtColor(image_cv, image_cv, COLOR_BGR2GRAY);
+
+        count++;
+
+        yInfo() <<"vgSLAMModule:writing images to buffers...";
```
Comment (traversaro) [inline doc/pipeline_thread/pipelineThread.dox]: `vgSLAMModule` seems to be a leftover

Comment (Nicogene) [reply doc/pipeline_thread/pipelineThread.dox]: Ops, yes it is 😅

**Comment by drdanz** [comment]
@Nicogene can we discuss this PR in person?

**Comment by Nicogene** [comment]
I would refactor the `PipelineThread` class using modern c++11 features.

**Pull request closed without merging**